docs/source/contributing_roles.rst

```diff
@@ -1,81 +1,227 @@
 
-    $ ansible-galaxy init --role-skeleton=contribute/_skeleton_role_ --init-path=roles ${NEWROLENAME}
+Therefore, roles must not reimplement functionality already present in other roles,
+or functionality that is not part of their intended purpose.
+Prerequisite tasks can be performed by roles added as dependencies, or by preceding roles
+in the playbook, and declared in the role documentation.
 
-Finally add a role documentation file at
-`docs/source/roles/role-${NEWROLENAME}.rst`. This file will need to contain
-a title and an include directive of the automatically generated documentation
-stub at the `../collections/osp/edpm/edpm_<NEWROLENAME>_role.rst` path.
+If a newly developed role has a prerequisite that isn't already fulfilled
+by any of the existing roles, and that is tangential to the primary purpose of the role,
+another role can be contributed, fulfilling this prerequisite.
 
-Optionally you can write further information about role operation.
-Including section of examples and molecule tests.
+All roles must have an `argument specification`_ metadata, for all of their endpoints,
+in the `roles/<NEWROLENAME>/meta/argument_specs.yml` file.
 
-Testing roles with molecule
-~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Argument specs allow for validity checks on arguments passed to the role in playbooks,
+while also providing bulk of the information neccessary for automated doc build.
+This both prevents unnecessary bugs, and simplifies documentation upkeep.
 
-Launch `make execute_molecule` to test all collections roles.
+Argument specification must contain definition of argument types, defaults and descriptions.
 
-Roles can be tested individually using the *molecule-venv* created by running
-`make execute_molecule`. The *molecule-venv* can also be manually created.
+All uses of the role, across playbooks and other roles, must utilize one of the endpoints
+defined in the argument specification metadata. No other variables than those defined
+as part of an endpoint may be used for role invocation.
 
-.. code-block:: console
+Properties of endpoints in argument specs are considered the source of truth.
+When they are in conflict with the actual implementation, for example when the default
+value of a variable in `roles/<NEWROLENAME>/defaults/main.yml` doesn't match
+the definition in metadata, the conflict must be resolved.
 
-    % python3 -m venv molecule-venv
-    % source molecule-venv/bin/activate
-    (molecule-venv) % pip install --upgrade pip
-    (molecule-venv) % pip install -r ../../molecule-requirements.txt
+Argument spec is essentially an API definition of the role. And as such must not be changed
+without regard to backward compatibility, or dependent software.
+Removal of variable from an argument spec must be preceded by a deprecation period.
+The same goes for a renaming of a variable, with a period of time when both new, and original
+names are accepted as part of the spec.
```

Comment:
This isn't really true right now. We have no reason to jump through hoops with a deprecation period when we haven't shipped anything.

Reply:
Our roles have a tight coupling with the operators that use it. The only deprecation period would be between PR's on edpm-ansible and the operator consuming the role, surely?

Reply:
We will allow partners to consume some of the roles. Especially the set of edpm_container* roles. We probably need a way to indicate which roles are for external consumption and which are not. The deprecation period across releases would apply to those that are intended for external consumption.

Reply:
I know it isn't the case right now. But this document is meant to guide long term development of this collection. We should at least make this the preferred practice if nothing else.